shell/platform/linux/public/flutter_linux/fl_engine.h

```diff
@@ -23,6 +24,99 @@ G_DECLARE_FINAL_TYPE(FlEngine, fl_engine, FL, ENGINE, GObject)
  * #FlEngine is an object that contains a running Flutter engine.
  */
 
+/**
+ * FlEnginePlatformMessageHandle:
+ *
+ * A handle used to respond to platform messages.
+ */
+typedef struct _FlEnginePlatformMessageHandle FlEnginePlatformMessageHandle;
+
+/**
+ * FlEnginePlatformMessageCallback:
+ * @engine: a #FlEngine
+ * @channel: channel message received on.
+ * @message: message content received from Dart
+ * @handle: a handle to respond to the message with
+ * @user_data: (closure): data provided when registering this callback
+ *
+ * Function called when platform messages are received.
+ */
+typedef void (*FlEnginePlatformMessageCallback)(
+    FlEngine* engine,
+    const gchar* channel,
+    GBytes* message,
+    FlEnginePlatformMessageHandle* handle,
+    gpointer user_data);
+
+/**
+ * fl_engine_set_platform_message_handler:
+ * @engine: a #FlEngine
+ * @channel: channel to listen on
+ * @callback: function to call when a message is received on this channel
+ * @user_data: (closure): user data to pass to @callback
+ *
+ * Register a callback to handle platform messages. Call
+ * fl_engine_send_platform_message_response() when this message should be
+ * responded to.
+ */
+void fl_engine_set_platform_message_handler(
+    FlEngine* engine,
+    const gchar* channel,
+    FlEnginePlatformMessageCallback callback,
+    gpointer user_data);
+
+/**
+ * fl_engine_send_platform_message_response:
+ * @engine: a #FlEngine
+ * @handle: message being responded to
+ * @response: (allow-none): response to send or %NULL for an empty response.
+ * @error: (allow-none): #GError location to store the error occurring, or %NULL
+ * to ignore.
+ *
+ * Respond to a platform message.
+ *
+ * Returns: %TRUE on success.
+ */
+gboolean fl_engine_send_platform_message_response(
+    FlEngine* engine,
+    FlEnginePlatformMessageHandle* handle,
+    GBytes* response,
+    GError** error);
+
+/**
+ * fl_engine_send_platform_message:
+ * @engine: a #FlEngine
+ * @channel: channel to send to
+ * @message: (allow-none): message buffer to send or %NULL for an empty message
+ * @cancellable: (allow-none): a #GCancellable or %NULL
+ * @callback: (scope async): a #GAsyncReadyCallback to call when the request is
+ * satisfied
+ * @user_data: (closure): user data to pass to @callback
+ *
+ * Asynchronously send a platform message.
+ */
+void fl_engine_send_platform_message(FlEngine* engine,
+                                     const gchar* channel,
+                                     GBytes* message,
+                                     GCancellable* cancellable,
+                                     GAsyncReadyCallback callback,
+                                     gpointer user_data);
+
+/**
+ * fl_engine_send_platform_message_finish:
+ * @engine: a #FlEngine
+ * @result: a #GAsyncResult
+ * @error: (allow-none): #GError location to store the error occurring, or %NULL
+ * to ignore.
+ *
+ * Complete request started with fl_engine_send_platform_message().
```

Comment:
Is there a nutshell description of what the purpose of this pattern is? I can see how you're using it in the example code, but how else would it ever be called (such that it's useful to have as a method rather than just providing the thing this would return directly to the callback)?

Reply:
This is the pattern used in GLib asynchronous code. The methods are normally called:

foo_async (object, args, cancellable, callback, user_data)
ReturnedData foo_sync (object, args, error)
ReturnedData foo_finish (object, result, error);

That should be recognisable to anyone familiar with GLib. In this case I dropped the _async suffix because we don't have a need for a _sync variant and it should be clear that the call is async. Other libraries do this.

Language bindings e.g. Vala hide away the three methods with their built in async/await operations. GLib doesn't use a Future type object for async.